langserver/handler.go

```diff
@@ -21,9 +21,33 @@ import (
 
 // NewHandler creates a Go language server handler.
 func NewHandler() jsonrpc2.Handler {
-	return jsonrpc2.AsyncHandler(jsonrpc2.HandlerWithError((&LangHandler{
+	return lspHandler{jsonrpc2.HandlerWithError((&LangHandler{
 		HandlerShared: &HandlerShared{},
-	}).handle))
+	}).handle)}
+}
+
+// lspHandler wraps LangHandler to correctly handle requests in the correct
+// order.
+//
+// The LSP spec dictates a strict ordering that requests should only be
+// processed serially in the order they are received. However, implementations
+// are allowed to do concurrent computation if it doesn't affect the
+// result. We actually can return responses out of order, since vscode does
+// not seem to have issues with that. We also do everything concurrently,
+// except methods which could mutate the state used by our typecheckers (ie
+// textDocument/didOpen, etc). Those are done serially since applying them out
+// of order could result in a different textDocument.
+type lspHandler struct {
+	jsonrpc2.Handler
+}
+
+// Handle implements jsonrpc2.Handler
+func (h lspHandler) Handle(ctx context.Context, conn *jsonrpc2.Conn, req *jsonrpc2.Request) {
+	if isFileSystemRequest(req.Method) {
```

Comment:
is this sufficient?

What if textDocument/references or other non-file system requests are processed before textDocument/didOpen or after textDocument/didClose?